gpjax/decision_making/decision_maker.py

```diff
@@ -0,0 +1,254 @@
+            observations throughout the decision making loop. In a typical setup they are
+            also used to fit the posteriors, using the `posterior_handlers`. Tags are used
+            to distinguish datasets, and correspond to tags in `posterior_handlers`.
+        acquisition_function_builder (AbstractAcquisitionFunctionBuilder): Object which
+            builds acquisition functions from posteriors and datasets, to decide where
+            to query next. In a typical Bayesian optimisation setup the point chosen to
+            be queried next is the point which maximizes the acquisition function.
+        acquisition_maximizer (AbstractAcquisitionMaximizer): Object which maximizes
+            acquisition functions over the search space.
+        key (KeyArray): JAX random key, used to generate random numbers.
+        post_ask (List[Callable]): List of functions to be executed after each ask step.
+        post_tell (List[Callable]): List of functions to be executed after each tell
+            step.
+    """
+
+    search_space: AbstractSearchSpace
+    posterior_handlers: Dict[str, PosteriorHandler]
+    datasets: Dict[str, Dataset]
+    acquisition_function_builder: AbstractAcquisitionFunctionBuilder
+    acquisition_maximizer: AbstractAcquisitionMaximizer
+    key: KeyArray
+    post_ask: List[Callable] = field(
+        default_factory=list
+    )  # Specific type is List[Callable[[DecisionMaker, Float[Array, ["1 D"]]], None]] but causes Beartype issues
+    post_tell: List[Callable] = field(
+        default_factory=list
+    )  # Specific type is List[Callable[[DecisionMaker], None]] but causes Beartype issues
+
+    @abstractmethod
+    def ask(self, key: KeyArray) -> Float[Array, "1 D"]:
+        """
+        In a typical decision making setup this will use the
+        `acquisition_function_builder` to form an acquisition function and then return
+        the point which maximizes the acquisition function using the
+        `acquisition_maximizer` as the point to be queried next.
+
+        Args:
+            key (KeyArray): JAX PRNG key for controlling random state.
+
+        Returns:
+            Float[Array, "1 D"]: Point to be queried next
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def tell(self, observation_datasets: Mapping[str, Dataset], key: KeyArray):
+        """
+        Tell decision maker about new observations. In a typical decision making setup
+        we will update the datasets and posteriors with the new observations.
+
+        Args:
+            observation_datasets (Mapping[str, Dataset]): Dictionary of datasets
+                containing new observations. Tags are used to distinguish datasets, and
+                correspond to tags in `posterior_handlers` in a typical setup.
+            key (KeyArray): JAX PRNG key for controlling random state.
+        """
+        raise NotImplementedError
+
+
+@dataclass
+class DecisionMaker(AbstractDecisionMaker):
```

Comment:
I wonder if the more abstract one should not have acquisition functions or optimizers in?

Reply:
perhaps then this class could be utility-based decision maker or something?

Reply:
as the difference between the two does seem to be mainly about having an acq to play with (and plot)

Reply:
We could then have a RS one subclassed from abstract without a need for defining an acq